doc/design/upgrade.md

```diff
@@ -0,0 +1,118 @@
+
+#### Freeze following attributes definition in VCH configuration and container VM configuration
+- VCH configuration: ID, Name, Version
+- Container VM serial port configuration.
+- Container VM log/debug file location.
+
+#### Configuration Version
+
+Both VCH and container configuration will have version and the value is same with what vic-machine version command shown. After upgrade, VCH configuration version should be updated to new version, but container VM configuration version will still be old one.
+
+New version's VCH will work with both old and new versions containers.
+
+#### Embed iso file version
+
+vic-machine need to identify the iso file version before and after upgrade. Two options here:
+- Add version into iso file Primary Volume Descriptor, and read back from vic-machine
+- Leverage iso file name including version, e.g. appliance-0.5.0.iso
+
+vic-machine should check iso file version during deployment, and after upload to datastore, version should be appended to iso file name, to make sure mutilple iso files could co-exist in the same VCH, and used for different version's container.
+
+Image file name should not be changed in datastore, cause the file path is used to create container VM. And then vic-machine will leverage file name for version checking during upgrade, to avoid download iso file from datastore back to where vic-machine is running.
+
+Note: No feasible golang library found for this function, so will write our own library to read iso metadata.
+
+### Configuration Data Migration
+
+To support VCH upgrade, any changes in VCH configuration and container VM configuration will introduce additional effort to migrate data from old version to newer version. And if we support a long upgrade path, for example from 2000->2500->3000->3500->4000,  there will have a list of utilies recognize old configuration structure to help data migration.
+
+#### Refactor current VCH configuration and container VM configuration structure
+To make data migration easier, we'll need to refactor current VCH and container VM configuration structure.
+
+VirtualContainerHostConfigSpec is too big, which includes everything in one structure. We're not be able to update part of configuration structure, it will help to update part of data with separated structures based on functionality, e.g. network, storage, and common attributes.
+
+For ContainerVM configuration, there are a few attributes not used for container VM setup, e.g. in ExecutorConfig, Key, LayerID, RepoName, are all not container VM configuration related. Move out irrelated attributes can help the structure stability in the future. 
+
+### Transactional Operation/Roll Back
+VM snapshot is one good option for upgrade roll back. Before upgrade, vic-machine will create one snapshot of VCH. If anything wrong happens, vic-machine can switch back to the pre-upgrade snapshot.
+
+As snapshot does not change iso files, vic-machine still need to switch the appliance iso files back to old one.
```

Comment:
No need to switch ISOs as the VMX is backed up by the snapshot. That means restoring the snapshot will restore the old path of the ISO.
We only need to have the ISOs have distinct names and to leave the old one in place.